pkg/addon/addon.go

```diff
@@ -1596,3 +1596,41 @@ func PackageAddon(addonDictPath string) (string, error) {
 	}
 	return archive, nil
 }
+
+// GetAddonLegacyParameters get addon's legacy parameters, that is stored in Secret
+func GetAddonLegacyParameters(ctx context.Context, k8sClient client.Client, addonName string) (map[string]interface{}, error) {
+	var sec v1.Secret
+	err := k8sClient.Get(ctx, client.ObjectKey{Namespace: types.DefaultKubeVelaNS, Name: addonutil.Addon2SecName(addonName)}, &sec)
+	if err != nil {
+		return nil, err
+	}
+	args, err := FetchArgsFromSecret(&sec)
+	if err != nil {
+		return nil, err
+	}
+	return args, nil
+}
+
+// MergeAddonInstallArgs merge addon's legacy parameter and new input args
+func MergeAddonInstallArgs(ctx context.Context, k8sClient client.Client, addonName string, args map[string]interface{}) (map[string]interface{}, error) {
+	legacyParams, err := GetAddonLegacyParameters(ctx, k8sClient, addonName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		return args, nil
+	}
+
+	if args == nil && legacyParams == nil {
+		return args, nil
+	}
+	r := make(map[string]interface{})
+	for k, v := range legacyParams {
+		r[k] = v
+	}
+	// override the legacyParams with new args for the same key
+	for k, v := range args {
+		r[k] = v
```

Comment:
What about nested arguments like this? It seems that this can only handle top-level arguments.

Old:
```json5
foo: {
  bar: "some"
  bar2: "some2"
}
```

New:
```json5
foo: {
  bar2: "some2new"
}
```

Now the old `foo.bar` will be gone, which is not an intended behavior.

Reply:
when user enable/upgrade addon, his command looks like this:
```
vela addon enable velaux --version=v1.4.5 repo=hub.vela.com  serviceType=NodePort imagePullSecrets={test-hub}
```
even "{test-hub}" looks like one object,  but it is the value of "imagePullSecrets" actually.
So I want to merge args, and overwrite the "value" without  recursively deep-merge .

Reply:
The `xxx=xxx` portion is a Helm strval. You can see it here:

https://github.com/kubevela/kubevela/blob/203a7dfbdab5f966d7085b21e090781395c79205/references/cli/addon.go#L317-L319

So the user can do something like this: `foo.bar=whatever` to define nested args.

An example is:

`vela addon enable chartmuseum persistence.enabled=true persistence.pvcName=myname`

Reply:
updated